### PR TITLE
Development of unit tests

### PR DESCRIPTION
**Pull request by Matheusstreisky**
Development of unit tests of controllers, services and repositories.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -75,6 +75,11 @@
 			<artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
 			<version>2.2.0</version>
 		</dependency>
+		<dependency>
+			<groupId>com.google.code.gson</groupId>
+			<artifactId>gson</artifactId>
```
Comment (JoaoPucci) [inline pom.xml:80]: If I am not mistaken, spring already has an embedded Jackson parser. Does gson has better performance or other benefits in comparison?

File: src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                                .contentType(MediaType.TEXT_PLAIN)
+                )
+                .andReturn().getResponse();
+
+        Assertions.assertEquals(HttpStatus.NO_CONTENT.value(), response.getStatus());
+    }
+
+    @Test
+    @WithMockUser
+    void requestFindTest() throws Exception {
```
Comment (JoaoPucci) [inline src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java]: Could add tests for corner cases, for example test the endpoint's response for when there are no results found.

File: src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        Mockito.when(productService.find(Mockito.anyLong())).thenReturn(PRODUCT_DTO_ACTIVE);
+
+        MockHttpServletResponse response = mockMvc
+                .perform(
+                        MockMvcRequestBuilders
+                                .get(PRODUCT_URL_WITH_ID, Mockito.anyLong())
```
Comment (JoaoPucci) [inline src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java]: Prefer avoiding "any" arguments in order to make the tests more assertive. For example, using a specific long value instead of `anyLong`, this makes sure that the correct parameter is received by the service class.

File: src/test/java/com/streisky/virtualshowcasebackend/domain/account/repository/AccountRepositoryTest.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import com.streisky.virtualshowcasebackend.domain.account.entity.Account;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.test.context.ActiveProfiles;
+
+@DataJpaTest
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
```
Comment (JoaoPucci) [inline src/test/java/com/streisky/virtualshowcasebackend/domain/account/repository/AccountRepositoryTest.java]: Question: Is this annotation needed? It doesn't get the test database configured on the properties file automatically?

Comment (JoaoPucci) [reply src/test/java/com/streisky/virtualshowcasebackend/domain/account/repository/AccountRepositoryTest.java]: Same question for `@ActiveProfiles("test")` annotations.

File: src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+                .andReturn().getResponse();
+        String jsonExpected = productDTOResponseJson.write(PRODUCT_DTO_ACTIVE).getJson();
+
+        Assertions.assertEquals(HttpStatus.CREATED.value(), response.getStatus());
+        Assertions.assertEquals(JsonParser.parseString(jsonExpected), JsonParser.parseString(response.getContentAsString()));
+    }
+
+    @Test
+    @WithMockUser
+    void requestPutWithInvalidBodyTest() throws Exception {
```
Comment (JoaoPucci) [inline src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java]: I think names could be a little more descriptive. For example, what part of the body is invalid, and what is the expected result?

There are some patterns for naming that use underscore (_) to better separate contexts.
I particularly like using `Should_ExpectedBehavior_When_StateUnderTest`, but that is up to discussion.

File: src/test/java/com/streisky/virtualshowcasebackend/domain/product/service/ProductServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        Assertions.assertEquals(productDTO, productEntity.toDTO());
+    }
+
+    @Test
+    public void updateSuccessfullyTest() {
+        ProductDTO productToUpdateDTO = new ProductDTO(1L, true, "Product updated", new BigDecimal(30),
+                List.of(
+                        new ImageDTO(1L, "Image updated test 1", new byte[1]),
+                        new ImageDTO(2L, "Image updated test 2", new byte[2])
+                ), "Test update");;
```
Comment (JoaoPucci) [inline src/test/java/com/streisky/virtualshowcasebackend/domain/product/service/ProductServiceTest.java]: 2 commas. Couldn't be more Java lol

File: pom.xml (hunk 1 above)
Comment (Matheusstreisky) [reply pom.xml:80]: Yes, spring already has. But, the implementation of Json parser of spring convert strings to List\<Object\> or Map\<String, Object\> according to the [JsonParser documentation](https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/json/JsonParser.html).

I wanted convert to JsonObject and compare the Json's, that's why I use Gson from Google.

File: src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java (hunk 3 above)
Comment (Matheusstreisky) [reply src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java]: How I do mock of return of the Service, can be any Long.

The correct would be not to mock the return of the service and use the value it returns?

File: src/test/java/com/streisky/virtualshowcasebackend/domain/account/repository/AccountRepositoryTest.java (hunk 4 above)
Comment (Matheusstreisky) [reply src/test/java/com/streisky/virtualshowcasebackend/domain/account/repository/AccountRepositoryTest.java]: The "AutoConfigureTestDatabase" is to use a real database and not be replaced by memory database. In beginning of tests implementation I was using a real database, but after I switch and forgot to remove this anotation. This is removed.

The "ActiveProfiles" is necessary to define a active profile in tests. Because this annotation, the spring acess application-test.properties intead of application.properties.

File: src/test/java/com/streisky/virtualshowcasebackend/domain/product/service/ProductServiceTest.java (hunk 6 above)
Comment (Matheusstreisky) [reply src/test/java/com/streisky/virtualshowcasebackend/domain/product/service/ProductServiceTest.java]: LOL, removed.

File: src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java (hunk 2 above)
Comment (Matheusstreisky) [reply src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java:125]: The case of results not found is tested in the resourceNotFoundTest() method. In this method I group all cases which some result is not found.

Comment (Matheusstreisky) [reply src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java]: I not do this cases because I mock return of the Service, and the Service methods I also tested.

The correct would be not to mock the return of the service and use the value it returns?

File: src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java (hunk 5 above)
Comment (Matheusstreisky) [reply src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java]: Renamed all tests methods.

File: src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java (hunk 2 above)
Comment (JoaoPucci) [reply src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java]: No, you can keep them mocked. But you would mock to the values for such corner cases, then you can test if the intended result code and JSON response is being delivered as expected, and the controller is not messing up anything on the middle.

On a second note, there is also the possibility of not mocking the services there and implementing integration tests for the controller layer. Not sure how this would work for Spring, though.

File: src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java (hunk 3 above)
Comment (JoaoPucci) [reply src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java]: You can just set up any specific long. The goal is just to make sure that the same value that is expected in the service call is being triggered by the controller.
Another thing that is nice to do is to take a sample of different possible values and writing different tests for them. For example, writing different tests for values like 7, -7, 0 or max long.

File: src/test/java/com/streisky/virtualshowcasebackend/domain/account/repository/AccountRepositoryTest.java (hunk 4 above)
Comment (JoaoPucci) [reply src/test/java/com/streisky/virtualshowcasebackend/domain/account/repository/AccountRepositoryTest.java]: IIRC, Spring should be able to get the test context automatically since these files are on the test sources. Maybe this is different for newer versions, but worth checking.

**Review by JoaoPucci — APPROVED** [review]

File: src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java (hunk 2 above)
Comment (Matheusstreisky) [reply src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java]: Hmm, got it.

Updated tests to use a fixed value.

File: src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java (hunk 3 above)
Comment (Matheusstreisky) [reply src/test/java/com/streisky/virtualshowcasebackend/controller/product/ProductControllerTest.java]: Hmm, got it.

Updated tests to use a fixed value

File: src/test/java/com/streisky/virtualshowcasebackend/domain/account/repository/AccountRepositoryTest.java (hunk 4 above)
Comment (Matheusstreisky) [reply src/test/java/com/streisky/virtualshowcasebackend/domain/account/repository/AccountRepositoryTest.java]: I not create a application.properties in the tests folder. I created in the resources folder of project. Because this is necessary define the active profile I guess. 
There are two different approaches I imagine, but I will keep this way as it is working correctly.
